source/grammar/qasm3Parser.g4

```diff
@@ -4,362 +4,210 @@ options {
+aliasDeclarationStatement: LET Identifier EQUALS aliasExpression SEMICOLON;
+classicalDeclarationStatement: (scalarType | arrayType) Identifier (EQUALS declarationExpression)? SEMICOLON;
+constDeclarationStatement: CONST scalarType Identifier EQUALS declarationExpression SEMICOLON;
+ioDeclarationStatement: (INPUT | OUTPUT) (scalarType | arrayType) Identifier SEMICOLON;
+oldStyleDeclarationStatement: (CREG | QREG) Identifier designator? SEMICOLON;
+quantumDeclarationStatement: qubitType Identifier SEMICOLON;
+
+// Declarations and definitions of higher-order objects.
+defStatement: DEF Identifier LPAREN argumentDefinitionList? RPAREN returnSignature? scope;
+externStatement: EXTERN Identifier LPAREN externArgumentList? RPAREN returnSignature? SEMICOLON;
+gateStatement: GATE Identifier (LPAREN params=identifierList? RPAREN)? qubits=identifierList scope;
+
+// Non-declaration assignments and calculations.
+assignmentStatement: indexedIdentifier op=(EQUALS | CompoundAssignmentOperator) expression SEMICOLON;
 expressionStatement: expression SEMICOLON;
 
-expression
-    // include terminator/unary as base cases to simplify parsing
-    : expressionTerminator
-    | unaryExpression
-    // expression hierarchy
-    | logicalAndExpression
-    | expression DOUBLE_PIPE logicalAndExpression
-    ;
-
-/**  Expression hierarchy for non-terminators. Adapted from ANTLR4 C
-  *  grammar: https://github.com/antlr/grammars-v4/blob/master/c/C.g4
-  * Order (first to last evaluation):
-    Terminator (including Parens),
-    Unary Op,
-    Multiplicative
-    Additive
-    Bit Shift
-    Comparison
-    Equality
-    Bit And
-    Exlusive Or (xOr)
-    Bit Or
-    Logical And
-    Logical Or
-**/
-
-logicalAndExpression
-    : bitOrExpression
-    | logicalAndExpression DOUBLE_AMPERSAND bitOrExpression
-    ;
-
-bitOrExpression
-    : xOrExpression
-    | bitOrExpression PIPE xOrExpression
-    ;
-
-xOrExpression
-    : bitAndExpression
-    | xOrExpression CARET bitAndExpression
-    ;
-
-bitAndExpression
-    : equalityExpression
-    | bitAndExpression AMPERSAND equalityExpression
-    ;
-
-equalityExpression
-    : comparisonExpression
-    | equalityExpression EqualityOperator comparisonExpression
-    ;
-
-comparisonExpression
-    : bitShiftExpression
-    | comparisonExpression ComparisonOperator bitShiftExpression
-    ;
-
-bitShiftExpression
-    : additiveExpression
-    | bitShiftExpression BitshiftOperator additiveExpression
-    ;
-
-additiveExpression
-    : multiplicativeExpression
-    | additiveExpression (PLUS | MINUS) multiplicativeExpression
-    ;
-
-multiplicativeExpression
-    // base case either terminator or unary
-    : unaryExpression
-    | multiplicativeExpression (ASTERISK | SLASH | PERCENT) unaryExpression
-    ;
-
-unaryExpression: unaryOperator? powerExpression;
+// TODO: this handling of the defcal block is incorrect, because it is not
+// constrained to consume balanced brace blocks.  This needs more attention.
```

Comment:
It's fine if all the input is valid, but the current rule allows ANTLR to "successfully" parse a token stream that should fail, provided the textual input can be tokenised.  For example,
```qasm
defcal x $0 {
}

{
2 3 4 5 6
}
```
will parse into a single `defcal` block with the current version - ANTLR will look ahead, fail to parse `2 3 4 5 6` into any valid rule if the `defcal` finishes on the first brace, so it will just eat tokens until those remaining can be parsed again (which in this case means eating up to the second brace).

It also has a problem that if the `defcalgrammar` contains tokens that aren't part of OQ3.0 (so the `qasm3Lexer.g4` lexer can't lex them), then this would all fail.  The more complete path forwards if we can make _no_ assumptions about the `defcalgrammar` language is to swap the lexer mode when we see the `defcal` token (like how we do for version numbers), and have it just eat every character between the opening brace and the closing brace into a single token.  We can count up the opening and closing braces to make sure we match the correct closing brace, subject to any "escape character" rules we'd want to implement.

The alternative approach is to hardcode a single grammar (OpenPulse) for the content of `defcal` blocks, and parse everything into a single AST.  I believe this is the more likely approach for _real_ implementations of the language, but for a reference implementation, we may need to be as true as possible to the full set of possible implementations.